warehouse/query-core/src/main/java/datawave/query/jexl/lookups/FieldExpansionIndexLookup.java

```diff
@@ -0,0 +1,168 @@
+package datawave.query.jexl.lookups;
+
+import java.util.HashSet;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.apache.accumulo.core.client.IteratorSetting;
+import org.apache.accumulo.core.client.Scanner;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.PartialKey;
+import org.apache.accumulo.core.data.Range;
+import org.apache.accumulo.core.data.Value;
+import org.apache.hadoop.io.Text;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+
+import datawave.core.iterators.FieldExpansionIterator;
+import datawave.query.config.ShardQueryConfiguration;
+import datawave.query.tables.ScannerFactory;
+import datawave.util.time.DateHelper;
+
+/**
+ * An {@link IndexLookup} that wraps {@link FieldExpansionIterator}.
+ */
+public class FieldExpansionIndexLookup extends AsyncIndexLookup {
+
+    private static final Logger log = LoggerFactory.getLogger(FieldExpansionIndexLookup.class);
+
+    protected String term;
+    protected Future<Boolean> timedScanFuture;
+    protected AtomicLong lookupStartTimeMillis = new AtomicLong(Long.MAX_VALUE);
+    protected CountDownLatch lookupStartedLatch;
+    protected CountDownLatch lookupStoppedLatch;
+
+    private Scanner scanner;
+
+    public FieldExpansionIndexLookup(ShardQueryConfiguration config, ScannerFactory scannerFactory, String term, Set<String> fields,
+                    ExecutorService execService) {
+        super(config, scannerFactory, true, execService);
+        this.term = term.toLowerCase();
+        this.fields = new HashSet<>();
+        if (fields != null) {
+            this.fields.addAll(fields);
+        }
+    }
+
+    @Override
+    public void submit() {
+        if (indexLookupMap == null) {
+            indexLookupMap = new IndexLookupMap(config.getMaxUnfieldedExpansionThreshold(), config.getMaxValueExpansionThreshold());
+
+            try {
+                scanner = scannerFactory.newSingleScanner(config.getIndexTableName(), config.getAuthorizations(), config.getQuery());
+
+                Range range = getScanRange();
+                scanner.setRange(range);
+
+                for (String field : fields) {
+                    scanner.fetchColumnFamily(new Text(field));
+                }
+
+                IteratorSetting setting = createIteratorSetting();
+                scanner.addScanIterator(setting);
+
+                timedScanFuture = execService.submit(createTimedCallable(scanner));
+            } catch (Exception e) {
+                log.error("Error expanding term into discrete fields", e);
+                throw new RuntimeException(e);
```

Comment:
cleanup the scanner

Reply:
Even if an exception is thrown in the catch block, the finally statement will still execute, thus closing the scanner.

Reply:
If you get an exception inside the future that is true, but if you get an exception in the above method after creating the scanner, I don't see that there is a finally that will catch anything. If the exception. None of the scanner methods have throws on them, but you are catching Exception, which covers a lot more than expected throwables.

Reply:
Taking a second pass through this. The `createTimedCallable` method that handles executing the scanner will close it via `scannerFactory.close(scanner)`. I think that's appropriate because otherwise the finally block might close the scanner before the executor service has had a chance to execute the timed callable. 

I thought I had a finally block in a local commit that never made it up but that not correct. 

Just for fun I closed the scanner in a finally block for regex and bounded range and several tests blew up. I'll add comments to each lookup that scanners should never be closed here.